### PR TITLE
Consolidate base URL composition in tests (#3008)

### PR DESCRIPTION
**Pull request by achave11**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: test/service/test_pagination.py
Comment (noah-aviel-dove) [inline test/service/test_pagination.py:190]:
```suggestion
        response = requests.get(self.base_url.set(path='/index/files',
                                                  args=third_request_params).url)
```

File: test/service/test_pagination.py
Comment (noah-aviel-dove) [inline test/service/test_pagination.py:226]:
```suggestion
                response = requests.get(self.base_url.set(path='/index/files',
                                                          params=params).url)
```

File: test/service/test_query_shortener.py
Comment (noah-aviel-dove) [inline test/service/test_query_shortener.py:38]: This fits on one line (barely); personally I think it looks better as you've written it but you may want to unwrap

File: test/service/test_repository_projects.py
Comment (noah-aviel-dove) [inline test/service/test_repository_projects.py:41]: Single quotes

File: test/service/test_repository_proxy.py
Comment (noah-aviel-dove) [inline test/service/test_repository_proxy.py:139]: https://github.com/DataBiosphere/azul/blob/develop/CONTRIBUTING.rst#14line-wrapping-and-indentation
```suggestion
                    azul_url = self.base_url.set(path=['repository', 'files', file_uuid],
                                                 args=dict(catalog=self.catalog, version=file_version))
```

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:355]: aligned indent
```suggestion
                "next": self.base_url.set(path='/index/files',
                                          args=dict(size=5,
                                                    search_after='cbb998ce-ddaf-34fa-e163-d14b399c6b34',
                                                    search_after_uid='meta%2332')).url,
```
alternatively:
```suggestion
                "next": self.base_url.set(
                    path='/index/files',
                    args=dict(size=5,
                              search_after='cbb998ce-ddaf-34fa-e163-d14b399c6b34',
                              search_after_uid='meta%2332')
                ).url,
```
point being the opening and closing parentheses for the call to `set` need to have the same indentation style.

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:491]: likewise

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2240]: again, parentheses are not aligned

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2252]: parens again

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2262]: parens again

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2293]: parens again

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2307]: parens

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2328]: parens

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2449]: Using `query_params` here is fine, but make sure that either `args` or `query_params` (they are synonymous) are used consistently.

File: test/service/test_response.py
Comment (noah-aviel-dove) [inline test/service/test_response.py:2484]: again, closing paren is not aligned

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
There is a pervasive inconsistency in the way that the fURL paths are set. In some cases, we use:

```
     base_url.set(path='/path/to/thing')
```

And other times:

``` 
     base_url.set(path=('path', 'to', 'thing'))
```

(Note that in the first form, the leading `/` is unnecessary)

I don't know why/when we should prefer one form over the other, or whether it was a conscious decision on your part regarding which to use. I feel this may deserve its own mention in the contributing guide. Using a single `/`-separated string is fewer characters, but I feel that the tuple-based approach is more resilient. Probably a question for @hannes-ucsc .

**Comment by noah-aviel-dove** [comment]
The `drop!` commit is unnecessary now that we've switched from Travis to GitHub Actions.

**Comment by achave11** [comment]
> There is a pervasive inconsistency in the way that the fURL paths are set. In some cases, we use:
> 
> ```
>      base_url.set(path='/path/to/thing')
> ```
> 
> And other times:
> 
> ```
>      base_url.set(path=('path', 'to', 'thing'))
> ```
> 
> (Note that in the first form, the leading `/` is unnecessary)
> 
> I don't know why/when we should prefer one form over the other, or whether it was a conscious decision on your part regarding which to use. I feel this may deserve its own mention in the contributing guide. Using a single `/`-separated string is fewer characters, but I feel that the tuple-based approach is more resilient. Probably a question for @hannes-ucsc .

[From Hannes on a previous review](https://github.com/DataBiosphere/azul/pull/2914#discussion_r629793500) in regards to …
```python3
url = furl(url=self.base_url, 
               path=('manifest', 'files'),
```
>The tuple form only helps if you're interpolating other variable values into the path. Static paths are better written as fixed strings.

We shouldn't see any `base_url.set(path=('path', 'to', 'thing'))` at least not in this PR.

File: test/service/test_manifest.py
Comment (achave11) [reply test/service/test_manifest.py:1683]: >In the previous section, we used an actual dictionary for the filters arg, but here we use a string literal. Is there a reason for doing so? 

In this case yes. It appears that it gets removed from the args if it is an empty dict, since we use this URL for asserting the url in the response, the `'{}'` must be present due to its presence in the response's URL. That being said, I removed it from the other sections since it isn't needed.

**Review by noah-aviel-dove — CHANGES_REQUESTED** [review]
Checklist is incomplete.

File: test/service/test_manifest_async.py
Comment (achave11) [reply test/service/test_manifest_async.py:204]: >It bothers me that this is more lines than the original.

Agreed, but overall there is a net reduction in lines. Which isn't bad considering some lines were added for legibility (like in this case).

**Review by noah-aviel-dove — APPROVED** [review]
Looks good.

File: test/service/test_manifest_async.py
Comment (achave11) [inline test/service/test_manifest_async.py:276]: @hannes-ucsc , I agreed with [Noah's advice](https://github.com/DataBiosphere/azul/pull/3011#discussion_r639310472), is it an acceptable change? It was done uniformly for similar invocations.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Conflicts. No bump.

**Comment by hannes-ucsc** [comment]
Hold off with re-review until base PR lands.

File: test/service/test_app_logging.py
Comment (hannes-ucsc) [inline test/service/test_app_logging.py:40]: Undo the inlining of the the `url` variable and then 
```suggestion
                        'host': url.netloc,
```

File: test/service/test_manifest_async.py
Comment (hannes-ucsc) [inline test/service/test_manifest_async.py:191]:
```suggestion

                    for i, expected_status in enumerate(3 * [301] + [302]):
                        response = requests.get(url.url, params=params, allow_redirects=False)
```

Comment (hannes-ucsc) [inline test/service/test_manifest_async.py:191]: Didn't you mean to set the query parameters via the URL?

File: test/service/test_manifest_async.py
Comment (hannes-ucsc) [inline test/service/test_manifest_async.py:272]:
```suggestion
        url = self.base_url.set(path='/fetch/manifest/files', args=self.params)
        response = requests.get(url.url)
```

File: test/service/test_manifest_async.py
Comment (hannes-ucsc) [inline test/service/test_manifest_async.py:285]: ditto

File: test/service/test_manifest_async.py
Comment (hannes-ucsc) [inline test/service/test_manifest_async.py:296]: ditto

File: test/service/test_manifest_async.py
Comment (hannes-ucsc) [inline test/service/test_manifest_async.py:305]: ditto

File: test/service/test_pagination.py
Comment (hannes-ucsc) [inline test/service/test_pagination.py:117]: ditto

File: test/service/test_pagination.py
Comment (hannes-ucsc) [inline test/service/test_pagination.py:132]: ditto

File: test/service/test_pagination.py
Comment (hannes-ucsc) [inline test/service/test_pagination.py:148]: ditto

Comment (hannes-ucsc) [reply test/service/test_pagination.py:148]: and so on

File: test/service/test_response.py
Comment (hannes-ucsc) [inline test/service/test_response.py:2548]: OK, it's exactly 80 but I think we should wrap anyways.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
No bump.

File: test/service/test_manifest_async.py
Comment (achave11) [reply test/service/test_manifest_async.py]: I did, I must have missed this one. Some usages were a bit harder to discern.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I recall that we agreed to try `str(url)` instead of `url.url`.

File: test/app_test_case.py
Comment (hannes-ucsc) [inline test/app_test_case.py:113]: There are still a million places that you touched where `.url` is used. So annoying!

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/app_test_case.py
Comment (achave11) [reply test/app_test_case.py]: As per our PL discussion, we only addressed variables of `furl` like `url` that made use of the `.url` property, furthermore …
>I recall that we agreed to try `str(url)` instead of `url.url`.

seemed to me as a literal directive to address `url.url` -> `str(url)` only.

While I appreciate your feedback and I am profoundly grateful for your mentorship, I think `So annoying!` was unnecessary.

Comment (hannes-ucsc) [reply test/app_test_case.py]: > As per our PL discussion, we only addressed variables of `furl` like `url` that made use of the `.url` property, furthermore …
> 
> > I recall that we agreed to try `str(url)` instead of `url.url`.
> 
> seemed to me as a literal directive to address `url.url` -> `str(url)` only.

That's not what I meant and, to be honest, a naíve misinterpretation. There is no point in changing only literal occurrences of `url.url` to `str(url)`. The only sensible interpretation of my code fragment is to replace any use of the `url` property on any instance of the `furl` class to instead calling the `str` built-in on that instance. Anything else would create a completely meaningless and inconsistent mix of approaches.

> While I appreciate your feedback and I am profoundly grateful for your mentorship, I think `So annoying!` was unnecessary.

You, Trevor and I will discuss this offline.

File: test/service/test_manifest_async.py
Comment (hannes-ucsc) [inline test/service/test_manifest_async.py:201]: Don't reuse the instance. Create a new one.

File: test/service/test_pagination.py
Comment (hannes-ucsc) [inline test/service/test_pagination.py:237]: Same

File: test/service/test_pagination.py
Comment (hannes-ucsc) [inline test/service/test_pagination.py:246]: Same

File: test/service/test_repository_proxy.py
Comment (hannes-ucsc) [inline test/service/test_repository_proxy.py:246]:
```suggestion
                                                              headers={'Location': str(dss_url_with_token),
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/service/test_repository_proxy.py
Comment (achave11) [reply test/service/test_repository_proxy.py]: @hannes-ucsc it seems like there are some `url.url` instances of `furl` in `integration_test.py`, should we also address them in this PR? I am thinking yes, but just want to confirm.

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**
